Add builtin Pyrefly language server support

grievejia opened the pull request:
I'm continuing the work of @lolpack's PR https://github.com/zed-industries/zed/pull/40953. This makes it such that user no longer needs to install the external pyrefly extension to use pyrefly. It also auto-downloads Pyrefly binary when there's no existing installation, which the pyrefly extension currently can't do. 

We don't publish Pyrefly release artifacts to github (yet), so the easiest way to obtain a Pyrefly installation is with PyPI. Given that the PyLSP adapter is already doing a PyPI-based installation, I've essentially copy&pasted most of the logic from there into pyrefly adapter so Pyrefly could work in exactly the same way for now. 

There are a lot of code duplications here! My thought was to get the functionality in first, and I can send followup PRs to extract duplicated code into common helper functions. 

I tested the change in 2 scenarios:
- No `pyrefly` binary is found in `$PATH`
- There's an existing `pyrefly` binary in `$PATH` -- I obtained it with `uv tool install pyrefly`. 
In both scenarios I ran the test by removing the Pyrefly exteranl plugin, start Zed, open a `.py` file with type errors in it, and verify that cmd-click (gotodef) works. 

Console output for scenario 1:
```
> cargo run
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 31.62s
     Running `target/debug/zed`
2025-12-04T20:58:07-08:00 INFO  [zed] ========== starting zed version 0.217.0+dev.65f328bd614282c1c370672106a9a279f8f22a3a, sha 65f328b ==========
2025-12-04T20:58:08-08:00 WARN  [zed::reliability] Minidump endpoint not set
2025-12-04T20:58:08-08:00 INFO  [extension_host] extensions updated. loading 2, reloading 0, unloading 0
2025-12-04T20:58:08-08:00 WARN  [audio::audio_settings] Denoise does not work on debug builds, not enabling
2025-12-04T20:58:08-08:00 INFO  [workspace] Rendered first frame
2025-12-04T20:58:09-08:00 WARN  [audio::audio_settings] Denoise does not work on debug builds, not enabling
2025-12-04T20:58:16-08:00 INFO  [lsp] starting language server process. binary path: "/Users/grievejia/Library/Application Support/Zed/languages/ruff/ruff-0.14.8/ruff-aarch64-apple-darwin/ruff", working directory: "/Users/grievejia/scratch/pyre_test", args: ["server"]
2025-12-04T20:58:20-08:00 INFO  [lsp] starting language server process. binary path: "/Users/grievejia/Library/Application Support/Zed/languages/pyrefly/pyrefly-venv/bin/pyrefly", working directory: "/Users/grievejia/scratch/pyre_test", args: ["lsp"]
...
```
We could see that the venv installation is successful and the auto-installed Pyrefly binary is used.

Console output for scenario 2:
```
> cargo run
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 29.62s
     Running `target/debug/zed`
2025-12-04T21:01:26-08:00 INFO  [zed] ========== starting zed version 0.217.0+dev.65f328bd614282c1c370672106a9a279f8f22a3a, sha 65f328b ==========
2025-12-04T21:01:27-08:00 WARN  [zed::reliability] Minidump endpoint not set
2025-12-04T21:01:27-08:00 INFO  [extension_host] extensions updated. loading 2, reloading 0, unloading 0
2025-12-04T21:01:27-08:00 WARN  [audio::audio_settings] Denoise does not work on debug builds, not enabling
2025-12-04T21:01:27-08:00 INFO  [workspace] Rendered first frame
2025-12-04T21:01:28-08:00 WARN  [audio::audio_settings] Denoise does not work on debug builds, not enabling
2025-12-04T21:01:34-08:00 INFO  [language] found user-installed language server for pyrefly. path: "/Users/grievejia/.local/bin/pyrefly", arguments: ["lsp"]
2025-12-04T21:01:34-08:00 INFO  [lsp] starting language server process. binary path: "/Users/grievejia/.local/bin/pyrefly", working directory: "/Users/grievejia/scratch/pyre_test", args: ["lsp"]
```
We could see that the correct installed Pyrefly binary was picked up. 


Release Notes:
- Added Pyrefly as builtin language server option

grievejia committed:
Add basic builtin Pyrefly language server support

grievejia committed:
Add documentation

osiewicz reviewed (commented):
Just my 2 cents on a first skim:
- You should probably implement `workspace_configuration` from `LspAdapter` trait, as that way you'll be able to propagate the toolchain that the user is using into pyrefly's configuration. You'd probably need to fill out `python-interpreter-path` (or whatever the setting is for pyrefly) with the provided `toolchain` value.
- I am a bit uneasy about installing pyrefly via PyPI. Yes, we went that route with pylsp because there was no other option; we've received quite a few bug reports to that effect ever since that's been rolled out (a year or so ago). Not that this is a show-stopper, but I'd mildly recommend considering installing the binary directly.

grievejia committed:
workplace config

grievejia committed:
format

osiewicz commented on crates/languages/src/python.rs:
```diff
@@ -2364,6 +2364,262 @@ impl LspInstaller for RuffLspAdapter {
+
+                if let Some(interpreter_path) = &env.environment.executable {
+                    object.insert(
+                        "pythonPath".to_owned(),
```
Is `pythonPath` a way to specify an active virtual environment with `pyrefly`? I can't find it in your docs (https://pyrefly.org/en/docs/configuration/)

grievejia committed:
Switch PyreflyLspAdapter from PyPI to GitHub binary releases

Replace the PyPI/venv-based installation mechanism for Pyrefly with direct GitHub binary release downloads, following the same pattern used by RuffLspAdapter and TyLspAdapter. This addresses reviewer concerns about the PyPI approach and takes advantage of Pyrefly v0.52.0+ publishing pre-built binaries on GitHub releases.

- Replace `python_venv_base: OnceCell` with `fs: Arc<dyn Fs>` on the struct
- Remove `ensure_venv`, `find_base_python`, and `base_venv` helper methods
- Add platform-specific constants for asset kind and OS name (macOS, Linux, Windows; no FreeBSD since Pyrefly doesn't publish builds for it)
- Add `build_asset_name()` using Pyrefly's `{name}-{os}-{arch}` naming convention, mapping `aarch64` to `arm64` to match their release assets
- Rewrite `LspInstaller` impl to use `GitHubLspBinaryVersion`, download from `facebook/pyrefly` GitHub releases, verify digests, and cache binaries
- Update `check_if_user_installed` to check venv path before PATH lookup, consistent with Ruff
- Pass `fs` to `PyreflyLspAdapter::new()` in lib.rs

grievejia committed:
Fix PyreflyLspAdapter to match upstream API changes

After rebasing onto upstream main, two API changes broke the Pyrefly adapter:

- `label_for_symbol` trait signature changed from separate `(name, kind)` parameters to a single `&Symbol` parameter. Updated to use `symbol.name` and `symbol.kind`.
- `cx.update()` in `workspace_configuration` no longer wraps the return value in `Result`. Restructured to match the Ty adapter pattern: extract settings from `cx.update()`, process toolchain data outside the closure, and explicitly return `Ok(user_settings)`.

To test:
- Enable Pyrefly in settings: `"language_servers": ["pyrefly", "..."]` under `languages.Python`
- Open a Python project and verify Pyrefly downloads from GitHub releases (check Zed log)
- Confirm go-to-definition, completions, and diagnostics work in `.py` files
- Restart Zed and verify cached binary is reused without re-downloading

osiewicz committed:
Merge branch 'main' into add_pyrefly_lsp

osiewicz committed:
Update to match latest main
